File: code/programs/go/unix-tools/comm_tool.go
// =========================================================================
// comm — Compare Two Sorted Files Line by Line
// =========================================================================
//
// The `comm` utility compares two sorted files and produces three columns
// of output:
//
//   Column 1: Lines unique to FILE1
//   Column 2: Lines unique to FILE2
//   Column 3: Lines common to both files
//
// # Visual example
//
//   FILE1:      FILE2:      comm FILE1 FILE2:
//   apple       banana      apple           ← only in FILE1
//   banana      cherry              banana  ← only in FILE2 (wait, no...)
//   cherry      date
//   fig
//
// Actually, let me show a proper example:
//
//   FILE1 (sorted):   FILE2 (sorted):   Output:
//   apple             banana            apple
//   banana            cherry                    banana
//   cherry                                      cherry
//   fig                                 fig
//
// Column 1 has no leading tabs.
// Column 2 has one leading tab.
// Column 3 has two leading tabs.
//
// # Suppression flags
//
//   -1  Suppress column 1 (lines unique to FILE1)
//   -2  Suppress column 2 (lines unique to FILE2)
//   -3  Suppress column 3 (lines common to both)
//
// Common combinations:
//   comm -12 FILE1 FILE2  → show only common lines
//   comm -23 FILE1 FILE2  → show only lines unique to FILE1
//   comm -13 FILE1 FILE2  → show only lines unique to FILE2
//
// # Architecture
//
//   comm.json (spec)             comm_tool.go (this file)
//   ┌──────────────────┐       ┌──────────────────────────────────┐
//   │ flags: -1,-2,-3   │       │ merge-compare two sorted lists   │
//   │ --output-delimiter│──────>│ classify each line into column   │
//   │ args: FILE1 FILE2 │       │ format with tab prefixes         │
//   └──────────────────┘       └──────────────────────────────────┘

package main

import (
	"bufio"
	"fmt"
	"io"
	"os"
	"strings"

	clibuilder "github.com/adhithyan15/coding-adventures/code/packages/go/cli-builder"
)

// =========================================================================
// CommOptions — configuration for the comm comparison
// =========================================================================

type CommOptions struct {
	SuppressCol1    bool   // -1: don't show lines unique to FILE1
	SuppressCol2    bool   // -2: don't show lines unique to FILE2
	SuppressCol3    bool   // -3: don't show lines common to both
	OutputDelimiter string // default: tab
}

// =========================================================================
// CommLine — a single line of comm output with its column assignment
// =========================================================================

type CommLine struct {
	Text   string // the actual line text
	Column int    // which column: 1, 2, or 3
}

// =========================================================================
// compareFiles — the core comm algorithm
// =========================================================================
//
// This implements a merge-compare on two sorted sequences, similar to
// the merge step of merge sort. We walk through both lists simultaneously:
//
//   - If line1 < line2: line1 is unique to FILE1 (column 1), advance FILE1
//   - If line1 > line2: line2 is unique to FILE2 (column 2), advance FILE2
//   - If line1 == line2: common line (column 3), advance both
//
// This is O(n + m) where n and m are the lengths of the two files.

func compareFiles(lines1, lines2 []string) []CommLine {
	var result []CommLine
	i, j := 0, 0

	for i < len(lines1) && j < len(lines2) {
		if lines1[i] < lines2[j] {
			// Line is unique to FILE1.
			result = append(result, CommLine{Text: lines1[i], Column: 1})
			i++
		} else if lines1[i] > lines2[j] {
			// Line is unique to FILE2.
			result = append(result, CommLine{Text: lines2[j], Column: 2})
			j++
		} else {
			// Line is common to both files.
			result = append(result, CommLine{Text: lines1[i], Column: 3})
			i++
			j++
		}
	}

	// Remaining lines in FILE1 are unique to FILE1.
	for ; i < len(lines1); i++ {
		result = append(result, CommLine{Text: lines1[i], Column: 1})
	}

	// Remaining lines in FILE2 are unique to FILE2.
	for ; j < len(lines2); j++ {
		result = append(result, CommLine{Text: lines2[j], Column: 2})
	}

	return result
}

// =========================================================================
// formatCommOutput — format comm results with tab prefixes
// =========================================================================
//
// Each column gets a prefix:
//   Column 1: no prefix
//   Column 2: one tab (or one delimiter)
//   Column 3: two tabs (or two delimiters)
//
// When columns are suppressed, the prefix width adjusts accordingly.
// For example, if column 1 is suppressed, column 2 gets no prefix and
// column 3 gets one tab.

func formatCommOutput(lines []CommLine, opts CommOptions) string {
	delim := opts.OutputDelimiter
	if delim == "" {
		delim = "\t"
	}

	var result strings.Builder

	for _, line := range lines {
		switch line.Column {
		case 1:
			if !opts.SuppressCol1 {
				result.WriteString(line.Text)
				result.WriteString("\n")
			}
		case 2:
			if !opts.SuppressCol2 {
				// Column 2 prefix: one delimiter for each non-suppressed
				// column before it.
				if !opts.SuppressCol1 {
					result.WriteString(delim)
				}
				result.WriteString(line.Text)
				result.WriteString("\n")
			}
		case 3:
			if !opts.SuppressCol3 {
				// Column 3 prefix: one delimiter for each non-suppressed
				// column before it.
				if !opts.SuppressCol1 {
					result.WriteString(delim)
				}
				if !opts.SuppressCol2 {
					result.WriteString(delim)
				}
				result.WriteString(line.Text)
				result.WriteString("\n")
			}
		}
	}

	return result.String()
}

// =========================================================================
// runComm — the testable core of the comm tool
// =========================================================================

func runComm(specPath string, argv []string, stdout io.Writer, stderr io.Writer) int {
	return runCommWithStdin(specPath, argv, stdout, stderr, os.Stdin)
}

// runCommWithStdin is the inner implementation with injectable stdin.

func runCommWithStdin(specPath string, argv []string, stdout io.Writer, stderr io.Writer, stdin io.Reader) int {
	// Step 1: Create the parser from the spec.
	parser, err := clibuilder.NewParser(specPath, argv)
	if err != nil {
		fmt.Fprintf(stderr, "comm: %s\n", err)
		return 1
	}

	// Step 2: Parse the arguments.
	result, err := parser.Parse()
	if err != nil {
		fmt.Fprintf(stderr, "%s\n", err)
		return 1
	}

	// Step 3: Handle the result.
	switch r := result.(type) {

	case *clibuilder.HelpResult:
		fmt.Fprintln(stdout, r.Text)
		return 0

	case *clibuilder.VersionResult:
		fmt.Fprintln(stdout, r.Version)
		return 0

	case *clibuilder.ParseResult:
		opts := CommOptions{
			SuppressCol1: getBool(r.Flags, "suppress_col1"),
			SuppressCol2: getBool(r.Flags, "suppress_col2"),
			SuppressCol3: getBool(r.Flags, "suppress_col3"),
		}
		if od, ok := r.Flags["output_delimiter"].(string); ok {
			opts.OutputDelimiter = od
		}

		// Read FILE1.
		file1Name, _ := r.Arguments["file1"].(string)
		file2Name, _ := r.Arguments["file2"].(string)

		lines1, err := readFileLines(file1Name, stdin)
		if err != nil {
			fmt.Fprintf(stderr, "comm: %s: %s\n", file1Name, err)
			return 1
		}

		lines2, err := readFileLines(file2Name, stdin)
		if err != nil {
			fmt.Fprintf(stderr, "comm: %s: %s\n", file2Name, err)
			return 1
		}

		// Compare and output.
		commLines := compareFiles(lines1, lines2)
		output := formatCommOutput(commLines, opts)
		fmt.Fprint(stdout, output)

		return 0

	default:
		fmt.Fprintf(stderr, "comm: unexpected result type: %T\n", result)
		return 1
	}
}

// =========================================================================
// readFileLines — read all lines from a file or stdin
// =========================================================================

func readFileLines(filename string, stdin io.Reader) ([]string, error) {
	var reader io.Reader
	if filename == "-" {
		reader = stdin
	} else {
		f, err := os.Open(filename)
		if err != nil {
			return nil, err
		}
		defer f.Close()
		reader = f
	}

	scanner := bufio.NewScanner(reader)
	var lines []string
	for scanner.Scan() {
		lines = append(lines, scanner.Text())
	}
	return lines, scanner.Err()
}

File: code/programs/rust/unix-tools/src/comm_tool.rs
//! # comm — Compare Two Sorted Files Line by Line
//!
//! This module implements the business logic for the `comm` command.
//! `comm` reads two sorted files and produces three columns of output:
//!
//! ```text
//!     Column 1: Lines unique to FILE1
//!     Column 2: Lines unique to FILE2
//!     Column 3: Lines common to both files
//! ```
//!
//! ## Visual Example
//!
//! ```text
//!     FILE1:    FILE2:    Output:
//!     apple     banana    apple            ← only in FILE1
//!     banana    cherry            banana   ← only in FILE2 (wait, banana is in both!)
//!     cherry    date              banana   ← common (column 3)
//!     fig       fig                       cherry  ← hmm, let's trace carefully...
//! ```
//!
//! Actually, let's trace the merge-like algorithm step by step:
//!
//! ```text
//!     FILE1:    FILE2:    Step:
//!     apple     banana    "apple" < "banana" → col1: "apple"
//!     banana    cherry    "banana" == "banana" → col3: "\t\tbanana"
//!     cherry    date      "cherry" == "cherry" → col3: "\t\tcherry"
//!     fig       fig       "date" < "fig" → col2: "\tdate"
//!                         "fig" == "fig" → col3: "\t\tfig"
//! ```
//!
//! ## Suppressing Columns
//!
//! The `-1`, `-2`, and `-3` flags suppress individual columns.
//! When a column is suppressed, its tab prefix is also removed,
//! so the remaining columns shift left.
//!
//! ## Requirement: Input Must Be Sorted
//!
//! `comm` assumes both files are sorted. If they aren't, the output
//! is undefined (and likely wrong). The `--check-order` flag makes
//! this an explicit error.

// ---------------------------------------------------------------------------
// Public API
// ---------------------------------------------------------------------------

/// Compare two sorted lists of lines and produce comm-style output.
///
/// # Parameters
///
/// - `lines1`: lines from FILE1, must be sorted
/// - `lines2`: lines from FILE2, must be sorted
/// - `suppress`: which columns to suppress: `[col1, col2, col3]`
/// - `output_delimiter`: separator between columns (default: TAB)
///
/// # Algorithm
///
/// This uses a merge-join algorithm, similar to merge sort's merge step.
/// We maintain two pointers (one into each file) and advance them
/// based on comparison:
///
/// ```text
///     if line1 < line2:  emit col1, advance ptr1
///     if line1 > line2:  emit col2, advance ptr2
///     if line1 == line2: emit col3, advance both
/// ```
///
/// This is O(n + m) where n and m are the lengths of the two files.
pub fn compare_sorted(
    lines1: &[String],
    lines2: &[String],
    suppress: [bool; 3],
    output_delimiter: &str,
) -> Vec<String> {
    let mut result = Vec::new();
    let mut i = 0;
    let mut j = 0;

    // The delimiter used for indentation. By default, TAB is used
    // to separate columns. Suppressed columns don't get a tab prefix.
    let sep = if output_delimiter.is_empty() { "\t" } else { output_delimiter };

    while i < lines1.len() && j < lines2.len() {
        match lines1[i].cmp(&lines2[j]) {
            std::cmp::Ordering::Less => {
                // Line is unique to FILE1 → column 1
                if !suppress[0] {
                    result.push(lines1[i].clone());
                }
                i += 1;
            }
            std::cmp::Ordering::Greater => {
                // Line is unique to FILE2 → column 2
                if !suppress[1] {
                    let prefix = if suppress[0] { "" } else { sep };
                    result.push(format!("{}{}", prefix, lines2[j]));
                }
                j += 1;
            }
            std::cmp::Ordering::Equal => {
                // Line is common to both → column 3
                if !suppress[2] {
                    let mut prefix = String::new();
                    if !suppress[0] { prefix.push_str(sep); }
                    if !suppress[1] { prefix.push_str(sep); }
                    result.push(format!("{}{}", prefix, lines1[i]));
                }
                i += 1;
                j += 1;
            }
        }
    }

    // --- Remaining lines from FILE1 ---
    while i < lines1.len() {
        if !suppress[0] {
            result.push(lines1[i].clone());
        }
        i += 1;
    }

    // --- Remaining lines from FILE2 ---
    while j < lines2.len() {
        if !suppress[1] {
            let prefix = if suppress[0] { "" } else { sep };
            result.push(format!("{}{}", prefix, lines2[j]));
        }
        j += 1;
    }

    result
}

// ---------------------------------------------------------------------------
// Unit tests
// ---------------------------------------------------------------------------

#[cfg(test)]
mod tests {
    use super::*;

    fn s(v: &[&str]) -> Vec<String> {
        v.iter().map(|s| s.to_string()).collect()
    }

    #[test]
    fn basic_three_columns() {
        let f1 = s(&["apple", "cherry", "fig"]);
        let f2 = s(&["banana", "cherry", "date"]);
        let result = compare_sorted(&f1, &f2, [false, false, false], "\t");
        assert_eq!(result, vec![
            "apple",
            "\tbanana",
            "\t\tcherry",
            "\tdate",
            "fig",
        ]);
    }

    #[test]
    fn suppress_col1() {
        let f1 = s(&["a", "b"]);
        let f2 = s(&["b", "c"]);
        let result = compare_sorted(&f1, &f2, [true, false, false], "\t");
        // Column 1 suppressed, so col2 shifts left
        assert_eq!(result, vec!["\tb", "c"]);
    }

    #[test]
    fn suppress_col2() {
        let f1 = s(&["a", "b"]);
        let f2 = s(&["b", "c"]);
        let result = compare_sorted(&f1, &f2, [false, true, false], "\t");
        assert_eq!(result, vec!["a", "\tb"]);
    }

    #[test]
    fn suppress_col3() {
        let f1 = s(&["a", "b"]);
        let f2 = s(&["b", "c"]);
        let result = compare_sorted(&f1, &f2, [false, false, true], "\t");
        assert_eq!(result, vec!["a", "\tc"]);
    }

    #[test]
    fn suppress_col1_and_col2() {
        let f1 = s(&["a", "b", "c"]);
        let f2 = s(&["b", "d"]);
        let result = compare_sorted(&f1, &f2, [true, true, false], "\t");
        // Only common lines remain
        assert_eq!(result, vec!["b"]);
    }

    #[test]
    fn empty_file1() {
        let f1: Vec<String> = Vec::new();
        let f2 = s(&["a", "b"]);
        let result = compare_sorted(&f1, &f2, [false, false, false], "\t");
        assert_eq!(result, vec!["\ta", "\tb"]);
    }

    #[test]
    fn empty_file2() {
        let f1 = s(&["a", "b"]);
        let f2: Vec<String> = Vec::new();
        let result = compare_sorted(&f1, &f2, [false, false, false], "\t");
        assert_eq!(result, vec!["a", "b"]);
    }

    #[test]
    fn both_empty() {
        let f1: Vec<String> = Vec::new();
        let f2: Vec<String> = Vec::new();
        let result = compare_sorted(&f1, &f2, [false, false, false], "\t");
        assert!(result.is_empty());
    }

    #[test]
    fn identical_files() {
        let f1 = s(&["a", "b", "c"]);
        let f2 = s(&["a", "b", "c"]);
        let result = compare_sorted(&f1, &f2, [false, false, false], "\t");
        assert_eq!(result, vec!["\t\ta", "\t\tb", "\t\tc"]);
    }

    #[test]
    fn no_common_lines() {
        let f1 = s(&["a", "c"]);
        let f2 = s(&["b", "d"]);
        let result = compare_sorted(&f1, &f2, [false, false, false], "\t");
        assert_eq!(result, vec!["a", "\tb", "c", "\td"]);
    }

    #[test]
    fn custom_output_delimiter() {
        let f1 = s(&["a", "b"]);
        let f2 = s(&["b", "c"]);
        let result = compare_sorted(&f1, &f2, [false, false, false], "|");
        assert_eq!(result, vec!["a", "||b", "|c"]);
    }
}
